### PR TITLE
fix: Ensure menus are always kept on the screen

### PR DESCRIPTION
**Pull request by MadMango**
completes #981 

still WIP, please see the comments on the files (I was gonna open it as a draft PR but forgot so just tagged it with WIP, soz)

### CODE REVIEW

**Comment by tommoor** [comment]
Hey hey, thanks for taking this on – really great stuff. I checked it out and noticed a couple of small issues with the way it works:

- If the menu is attached to a fix item such as the "New doc" button and the page is scrolled then it's placed in the wrong position:
    image

- It seems as though the opening animation is applied to the wrong element, the content of the dropdown moves within the menu bounding.
- We've also lost the subtle border radius on the dropdown, this might be related to the issue above though.

File: app/components/DropdownMenu/DropdownMenu.js [hunk 1]
```diff
@@ -26,9 +26,19 @@ type Props = {
...
+  @observable right: ?number;
+  @observable left: ?number;
+  @observable position: 'left' | 'right' | 'center';
+  @observable bodyRect: ClientRect;
+  @observable labelRect: ClientRect;
+  @observable dropdownRef: { current: null | HTMLElement };
+
+  constructor(props: Props) {
+    super(props);
+    this.dropdownRef = React.createRef();
```
Comment (tommoor) [inline app/components/DropdownMenu/DropdownMenu.js]: There's no need for the constructor here, you can actually just do this on the class itself

```javascript
dropdownRef: React.createRef();
```

File: app/components/DropdownMenu/DropdownMenu.js [hunk 2]
```diff
@@ -62,13 +67,65 @@ class DropdownMenu extends React.Component<Props> {
...
+          this.left = undefined;
+        } else if (this.position === 'left') {
+          this.position = 'right';
+          this.right = undefined;
+        }
+      }
+    }
+
+    this.initPosition();
+    this.forceUpdate();
```
[the remark this answers is not in the document]
Comment (tommoor) [reply app/components/DropdownMenu/DropdownMenu.js]: Updating an observable prop should just magically always re-render so I'm not sure either with this one 🤔

File: app/components/DropdownMenu/DropdownMenu.js (hunk 1 above)
Comment (MadMango) [reply app/components/DropdownMenu/DropdownMenu.js]: fixed, thanks!

**Comment by MadMango** [comment]
> Hey hey, thanks for taking this on – really great stuff. I checked it out and noticed a couple of small issues with the way it works:
> 
> * If the menu is attached to a fix item such as the "New doc" button and the page is scrolled then it's placed in the wrong position:
>   image
> * It seems as though the opening animation is applied to the wrong element, the content of the dropdown moves within the menu bounding.
> * We've also lost the subtle border radius on the dropdown, this might be related to the issue above though.

I started looking into this and I noticed a couple issues with changing the position to `fixed` and I changed back to `absolute` and I'm now using the scroll position of the window to determine the `bottom` position for the dropdown. However, the menu that you mentioned has the same behavior in master so I think that might be a separate issue? I'm happy to pick it up as well but not sure if it's better to track it as a separate problem or not.

**Comment by tommoor** [comment]
> However, the menu that you mentioned has the same behavior in master so I think that might be a separate issue?

Oh, are you sure? I just tried reproducing on the production website and I was unable to – just opening that menu on the dashboard and then scrolling the page down…

**Comment by MadMango** [comment]
> > However, the menu that you mentioned has the same behavior in master so I think that might be a separate issue?
> 
> Oh, are you sure? I just tried reproducing on the production website and I was unable to – just opening that menu on the dashboard and then scrolling the page down…

Hi Tom, sorry for very inconsistent replies but I've been away for most of the last two weeks, I'll look into it properly tomorrow.

**Comment by tommoor** [comment]
No worries, appreciate your help with this – hoping we can get it squared away for the next release 😄

**Comment by MadMango** [comment]
Hi Tom!

I'm not sure if we're trying the same thing but for me, the behaviour in master and the feature branch is exactly the same for that menu, unless it's something different / I'm using it wrong?

[master](https://drive.google.com/open?id=1zYrb6Z49iLdDVcGl9UrVz-IpOLP_RLbJ)
[feature](https://drive.google.com/open?id=1_OFZWVrPDwxWAv0jwzBSsmutCX9kDt37)

Thanks,
Mat

**Comment by tommoor** [comment]
The issue only shows up when the page is **scrolled** – because the menu takes into account the page scroll but in this case the menu is actually relative to a fixed element. I think the solution would be to find if any parent elements are fixed and then ignore the scroll?

**Comment by MadMango** [comment]
> The issue only shows up when the page is **scrolled** – because the menu takes into account the page scroll but in this case the menu is actually relative to a fixed element. I think the solution would be to find if any parent elements are fixed and then ignore the scroll?

I think for dropdowns that are children of fixed elements we could simply make the dropdown fixed as well, I just tested that and it seems to work fine, can you foresee any problems with that approach?

**Comment by MadMango** [comment]
@tommoor `getComputedStyle` saves us a lot of hassle here, I pushed a change, looks good to me now.

File: app/components/DropdownMenu/DropdownMenu.js [hunk 3]
```diff
@@ -40,18 +46,24 @@ class DropdownMenu extends React.Component<Props> {
...
+        this.bottom = undefined;
+        this.position = this.props.position || 'left';
+
+        if (currentTarget.parentElement) {
+          const triggerParentStyle = getComputedStyle(
+            currentTarget.parentElement
+          );
+
+          if (triggerParentStyle.position === 'static') {
+            this.fixed = true;
```
Comment (tommoor) [inline app/components/DropdownMenu/DropdownMenu.js]:
```suggestion
            this.fixed = true;
            this.top = this.labelRect.bottom;
```

Hey @MadMango – testing this out there still seems to be an issue with the positioning of fixed menus when the body is scrolled – but adding this additional line fixes it for me

**Pull request merged**